File: erlang/src/step1_read_print.erl
%%%
%%% Step 1: read/print
%%%

-module(step1_read_print).

-export([main/1]).

main(_) ->
    case io:get_line(standard_io, "user> ") of
        eof ->
            % break out of the loop
            io:format("~n"),
            ok;
        {error, Reason} ->
            io:format("Error reading input: ~s~n", [Reason]),
            exit(ioerr);
        Line ->
            print(eval(read(string:strip(Line, both, $\n)))),
            main("")
    end.

read(String) ->
    case reader:read_str(String) of
        {ok, Value} -> Value;
        {error, Reason} -> io:format("error: ~s~n", [Reason]), nil
    end.

eval(Value) ->
    Value.

print(none) ->
    % if nothing meaningful was entered, print nothing at all
    ok;
print(Value) ->
    io:format("~s~n", [printer:pr_str(Value, true)]).

File: elixir/lib/mix/tasks/step1_read_print.ex
defmodule Mix.Tasks.Step1ReadPrint do
  def run(_), do: loop

  defp loop do
    Mal.Core.readline("user> ")
      |> read_eval_print
      |> IO.puts

    loop
  end

  defp read(input) do
    Mal.Reader.read_str(input)
  end

  defp eval(ast), do: ast

  defp print(value) do
    Mal.Printer.print_str(value)
  end

  defp read_eval_print(:eof), do: exit(:normal)
  defp read_eval_print(line) do
    read(line)
      |> eval
      |> print
  catch
    {:error, message} -> IO.puts("Error: #{message}")
  end
end
